### ISSUE TITLE
Solve partial instantiation of generic methods in a clean way

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#32340]
There are a number of issues with the way the new partial instantiation of generic methods (a test of which can be found [language_2/partial_tearoff_instantiation_test](https://github.com/dart-lang/sdk/blob/master/tests/language_2/partial_tearoff_instantiation_test.dart)) was implemented in 42a639973dfd1ebe2de2f2b8b7b4c027986f1805 (as well as ac7f5800059659462b2a795c3a16d03a20aa6a2d which duplicated the problem into the constant evaluator).

**Issue 1)** If the frontend infers that there are only `dynamic` type arguments for the partial instantiation of the generic method, the VM will explicitly use a **non-canonicalized** TypeArguments vector.

-> See [StreamingFlowGraphBuilder::BuildPartialTearoffInstantiation](https://github.com/dart-lang/sdk/blob/ac7f5800059659462b2a795c3a16d03a20aa6a2d/runtime/vm/compiler/frontend/kernel_binary_flowgraph.cc#L8110)

  * Firstly this will cause us having the same TypeArguments vector of all-dynamic **many times in memory**
  * Secondly this actually breaks code in debug mode, where we have assertions (like [this one](https://github.com/dart-lang/sdk/blob/ac7f5800059659462b2a795c3a16d03a20aa6a2d/runtime/vm/runtime_entry.cc#L500)) which assert that type arguments are canonicalized

**Issue 2)** Whenever we want to find out if a `TypeArguments` vector was supplied by a caller or is available in a partially-instantiated closure, we cannot compare against `null`, because currently the VM uses `null` as a TypeArgument vector of all-dynamic.

-> For finding out whether the caller supplied function type arguments we have to look into the arguments descriptor.
-> For finding out whether the closure-at-hand is actually partially-instantiated and has therefore function type arguments, we have to use a mechanism which doesn't rely on `null` comparisons.

**Issue 3)** The `Closure::function_type_arguments` field can apparently not be used as a slot for storing the type arguments at a partial-instantiation. This field is used for a different purpose, @crelier should be able to elaborate on this.

-> Currently the partial instantiation implementation copies over all the fields from the non-instantiated closure, and overrides the `Closure::function_type_arguments` field with the type arguments supplied at partial-instantiation time. See [the code here](https://github.com/dart-lang/sdk/blob/ac7f5800059659462b2a795c3a16d03a20aa6a2d/runtime/vm/compiler/frontend/kernel_binary_flowgraph.cc#L8123).

/cc @crelier @kmillikin

**Comment by mkustermann** [comment]
If I have time, I will look into it in the near future.

**Comment by sjindel-google** [comment]
Do issues 2 and 3 cause any bugs?

**Comment by mkustermann** [comment]
To the best of my knowledge all 3 of these are bugs that need to be fixed (which is why I filed this issue)

**Comment by sjindel-google** [comment]
Do you have any reproductions to help?

**Comment by mkustermann** [comment]
> ... So I don't see a problem here.

As explained in the bug report and offline, I see an issue here, I don't think what you wrote is correct!

The 3 issues I've pointed out are real issues and need to be fixed!

**Comment by mkustermann** [comment]
To make this even clearer, here's a simple test case:

```dart
main() {
  print(foo<double>(1.2));
  print(partiallyInstantiatedClosure(42));
}

typedef String F(int f);
F partiallyInstantiatedClosure;

String foo<A>(A a) {
  String bar<B>(B b) {
    String hest<C>(C c) {
      return '$A:$a, $B:$b, $C:$c';
    }
    partiallyInstantiatedClosure = hest;
    return hest<int>(42);
  }
  return bar<String>('str');
}
```

This results in
```
% out/ReleaseX64/dart --preview_dart_2 test.dart
double:1.2, String:str, int:42
int:1.2, dynamic:str, dynamic:42

```
instead of
```
% out/ReleaseX64/dart --preview_dart_2 test.dart
double:1.2, String:str, int:42
double:1.2, String:str, int:42
```

**Quite obviously the TypeArguments vector is wrong!!!**

For reference, the kernel is:
```
  typedef F = (core::int) → core::String;
  static field (core::int) → core::String partiallyInstantiatedClosure;
  static method main() → dynamic {
    core::print(x::foo<core::double>(1.2));
    core::print(x::partiallyInstantiatedClosure.call(42));
  }
  static method foo<A extends core::Object>(x::foo::A a) → core::String {
    function bar<B extends core::Object>(B b) → core::String {
      function hest<C extends core::Object>(C c) → core::String {
        return "${x::foo::A}:${a}, ${B}:${b}, ${C}:${c}";
      }
      x::partiallyInstantiatedClosure = hest<core::int>;
      return hest.call<core::int>(42);
    }
    return bar.call<core::String>("str");
  }
```

**Comment by sjindel-google** [comment]
Thanks for the reproduction, it's very helpful.

I've talked with @eernstg , @crelier and @leafpetersen and discovered that this feature is actually not specified at all. My understanding was that type inference in the FE would only apply this operator to top-level and class-level tearoffs, not to local functions. #31665 explains how the spec. suggests this intepretation. "Issue 3", the overloading of `Closure::function_type_arguments`, was not a problem under this assumption.

The language team decided that there would be some restrictions on this feature similar to that, but they are slightly different, and do allow the case you've demonstrated above. I'm still not sure which restrictions, if any, the CFE is imposing during type inference. DDC is currently not imposing any restrictions at all.

To solve "Issue 3", the issue you have reproduced, we need to redesign the implementation. Obviously, the language team needs to specify this feature first.  

I've created a task for this issue (#32533), and a task for the discrepancy between the CFE's implementation and the VM's implementation (#32534).

As mentioned above, I've already created a P2-level task (#32527) for "Issue 1" and "Issue 2", which are bugs regardless of how the feature is specified.
